flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTimeoutTest.java

```diff
@@ -0,0 +1,200 @@
+
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static org.junit.jupiter.api.Assertions.fail;
+
+/** Test for timeout functionalities of {@link AsyncSinkWriter}. */
+public class AsyncSinkWriterTimeoutTest {
+    private final List<Long> destination = new ArrayList<>();
+
+    @Test
+    void writerShouldNotRetryIfRequestIsProcessedBeforeTimeout() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, false);
+        tpts.setCurrentTime(0);
+        writer.write("1", null);
+        tpts.setCurrentTime(10);
+        writer.deliverMessage();
+        tpts.setCurrentTime(120);
+        writer.flush(false);
+        assertThat(destination).containsExactly(1L);
+    }
+
+    @Test
+    void writerShouldRetryOnTimeoutIfFailOnErrorIsFalse() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, false);
+        tpts.setCurrentTime(0);
+        writer.write("1", null);
+        // element should be requeued back after timeout
+        tpts.setCurrentTime(110);
+        // deliver initial message after timeout
+        writer.deliverMessage();
+        // flush outstanding mailbox messages containing resubmission of the element
+        writer.flush(false);
+        assertThat(destination).containsExactly(1L, 1L);
+    }
+
+    @Test
+    void writerShouldFailOnTimeoutIfFailOnErrorIsTrue() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, true);
+        tpts.setCurrentTime(0);
+        writer.write("1", null);
+        tpts.setCurrentTime(110);
+        writer.deliverMessage();
+        assertThatExceptionOfType(FlinkRuntimeException.class)
+                .isThrownBy(() -> writer.flush(false))
+                .withCauseInstanceOf(TimeoutException.class)
+                .havingCause()
+                .withMessage("Request timed out after 100ms with failOnTimeout set to true.");
+    }
+
+    @Test
+    void writerShouldDiscardRetriedEntriesOnTimeout() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, false);
+        writer.setShouldFailRequest(true);
```

Comment:
Why dont you parametrize `shouldFailRequest` for each test? In this case, you might need to clean up the state of `shouldfailRequest` manually for each test no? (e.g., for this test there is no cleanup)

Reply:
I am not sure I follow, could you explain more?

Reply:
Not an issue anymore